src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -59,18 +61,60 @@ def generate_args(mypy: MyPy, *, file_list_path: str) -> Tuple[str, ...]:
     return tuple(args)
 
 
+# MyPy searches for types for a package in packages containing a `py.types` marker file or else in
+# a sibling `<package>-stubs` package as per PEP-0561. Going further than that PEP, MyPy restricts
+# its search to `site-packages`. Since PEX deliberately isolates itself from `site-packages` as
+# part of its raison d'être, we monkey-patch `site.getsitepackages` to look inside the scrubbed
+# PEX sys.path before handing off to `mypy`.
+#
+# As a complication, MyPy does its own validation to ensure packages aren't both available in
+# site-packages and on the PYTHONPATH. As such, we elide all PYTHONPATH entries from artificial
+# site-packages we set up since MyPy will manually scan PYTHONPATH outside this PEX to find
+# packages. We also elide the values of PEX_EXTRA_SYS_PATH, which will be relative paths unlike
+# every other entry of sys.path. (We can't directly look for PEX_EXTRA_SYS_PATH because Pex scrubs
+# it.)
+#
+# See:
+#   https://mypy.readthedocs.io/en/stable/installed_packages.html#installed-packages
+#   https://www.python.org/dev/peps/pep-0561/#stub-only-packages
+#   https://github.com/python/mypy/blob/f743b0af0f62ce4cf8612829e50310eb0a019724/mypy/sitepkgs.py#L22-L28
```

Comment:
This comment and following logic is almost all from John. The only part I added was the end about `PEX_EXTRA_SYS_PATH`.